src/algebra/lie/character.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.lie.abelian
+import algebra.lie.solvable
+import linear_algebra.dual
+
+/-!
+# Characters of Lie algebras
+
+A character of a Lie algebra is a morphism `L → R`. For an Abelian Lie algebra (e.g., a Cartan
+subalgebra of a semisimple Lie algebra) a character is just a linear form.
+
+## Main definitions
+
+  * `lie_algebra.lie_character`
+  * `lie_algebra.lie_character_equiv_linear_dual_of_abelian`
+
+## Tags
+
+lie algebra, lie character
+-/
+
+universes u v w w₁
+
+namespace lie_algebra
+
+variables (R : Type u) (L : Type v) [comm_ring R] [lie_ring L] [lie_algebra R L]
+
+/-- A character of a Lie algebra is a morphism to the scalars. -/
+abbreviation lie_character := L →ₗ⁅R⁆ R
+
+variables {R L}
+
+lemma lie_character_of_lie_eq_zero (χ : lie_character R L) (x y : L) : χ ⁅x, y⁆ = 0 :=
+by rw [lie_hom.map_lie, lie_ring.of_associative_ring_bracket, mul_comm, sub_self]
+
+lemma lie_character_mem_derived_eq_zero
+  (χ : lie_character R L) {x : L} (h : x ∈ derived_series R L 1) : χ x = 0 :=
+begin
+  rw [derived_series_def, derived_series_of_ideal_succ, derived_series_of_ideal_zero,
+    ← lie_submodule.mem_coe_submodule, lie_submodule.lie_ideal_oper_eq_linear_span] at h,
+  apply submodule.span_induction h,
+  { rintros y ⟨⟨z, hz⟩, ⟨⟨w, hw⟩, rfl⟩⟩, apply lie_character_of_lie_eq_zero, },
+  { exact χ.map_zero, },
+  { intros y z hy hz, rw [lie_hom.map_add, hy, hz, add_zero], },
+  { intros t y hy, rw [lie_hom.map_smul, hy, smul_zero], },
+end
+
+/-- For an Abelian Lie algebra, characters are just linear forms. -/
+def lie_character_equiv_linear_dual_of_abelian [is_lie_abelian L] :
+  lie_character R L ≃ module.dual R L :=
```

Comment:
This map is presumably R-linear, right?

Reply:
Yes except I haven't defined any structure on `lie_character R L` yet.

In general we don't have a module structure on Lie algebra morphisms. However for characters we do because they're just the same as a linear map from `L → R` that vanishes on `⁅⊤, ⊤⁆`.

This stuff does need to be added but I was planning on leaving it out of this PR to make it a bit more manageable. Of course I can certainly include it if you think I should.

Reply:
Thinking about this a little more, these results should probably be set up for morphisms into any Abelian Lie algebra.